bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/handler/DLinkSmartPlugHandler.java

```diff
@@ -0,0 +1,409 @@
+
+    private State getTotalConsumption() {
+        return getQuantity("GetPMWarningThreshold", "TotalConsumption", "2", SmartHomeUnits.KILOWATT_HOUR);
+    }
+
+    private State getTemperature() {
+        return getQuantity("GetCurrentTemperature", "CurrentTemperature", "3", SIUnits.CELSIUS);
+    }
+
+    private State getState() {
+        try {
+            String state = soapAction("GetSocketSettings", "OPStatus", moduleParameters("1")).toLowerCase();
+            if ("true".contentEquals(state)) {
+                return OnOffType.ON;
+            } else if ("false".equals(state)) {
+                return OnOffType.OFF;
+            } else {
+                logger.error("Unexpected state: {}", state);
+            }
+        } catch (CommunicationException e) {
+            logger.error("Failed to get state (communication error).");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+        }
+        return UnDefType.UNDEF;
+    }
+
+    private void setState(final OnOffType state) {
+        try {
+            final String encodedState = OnOffType.ON.equals(state) ? "true" : "false";
+            final String controlParameters = moduleParameters("1")
+                    + "<NickName>Socket 1</NickName><Description>Socket 1</Description>" + "<OPStatus>" + encodedState
+                    + "</OPStatus>";
+            final String response = soapAction("SetSocketSettings", "SetSocketSettingsResult", controlParameters);
+            if (!"ok".equals(response.toLowerCase())) {
+                logger.error("Failed to set state.");
+            }
+        } catch (Exception e) {
+            // pass
+        }
+    }
+
+    private String moduleParameters(final String module) {
+        return "<ModuleID>" + module + "</ModuleID>";
+    }
+
+    private State getQuantity(final String action, final String name, final String module, final Unit<?> unit) {
+        try {
+            final String s = soapAction(action, name, moduleParameters(module));
+            try {
+                if (s == null) {
+                    return UnDefType.UNDEF;
+                } else {
+                    return new QuantityType<>(Double.valueOf(s), unit);
+                }
+            } catch (NumberFormatException e) {
+                logger.error("Unexpected number format: {}", s);
+            }
+        } catch (CommunicationException e) {
+            logger.error("Failed to get quantity (communication error).");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+        }
+        return UnDefType.UNDEF;
+    }
+
+    private String soapAction(final String action, final String name, final String params)
+            throws CommunicationException {
+        Document document = post(action, requestBody(action, params), authentication);
+        return getElementByTagName(document, name);
+    }
+
+    private boolean authenticate() {
+        if (invalidConfiguration) {
```

Comment:
I still think you should attempt to use the existing communication class to avoid duplicate code.

Most of this method can be replaced with a call to `login` and `getHNAPStatus` from `DLinkHNAPCommunication`

Reply:
I don't want to offend you, but I think the existing `DLinkHNAPCommunication` is quite a mess. I had a hard time understanding the control flow. Maybe because I'm C++ and not Java developer, but I find that some of the patterns in the class are at least weird, if not confusing.
I find that the way how it is used in `DLinkMotionSensor` is even worse. Look at the `run()`-method of the `detect`-poller alone!
And I have not yet succeeded to use your `DLinkHNAPCommunication` class to set the state.

I think we have some options
1. You (and the other reviewer) accept the code duplication. I don't like it, too. But it would be the quickest and most pragmatic way to add support for the DSP-W215-Thing to openHAB.
2. You close the PR. I really don't like that option, but it runs perfectly for me at home, so I could live with it.
3. I change `DLinkSmartPlugHandler` to use `DLinkHNAPCommunication`. This won't happen in the near future.
4. You change `DLinkSmartPlugHandler` to use `DLinkHNAPCommunication`. Go ahead. If it works in the end, I'm fine.
5. I factor out the `HNAP`-stuff from `DLinkSmartPlugHandler` and make the `DLinkMotionSensor*` use that, too. I could do this, but I cannot test it.
6. I factor out the `HNAP`-stuff from `DLinkSmartPlugHandler `and you make the `DLinkMotionSensor*` use that, too.

My favorite solution is (6), but I understand if you have a different opinion.

Reply:
Have you actually tried using `DLinkHNAPCommunication` again since I pointed out the possible missing call to `setAuthenticationHeaders`?

It's a shame you didn't ask for help previously with `DLinkHNAPCommunication` because now it sounds like you've got to the point where you aren't willing to try using it. I agree with hindsight `DLinkHNAPCommunication` could be better, but we could make some changes to it to make it easier to use rather than going to the extreme of completely rewriting it.

I'll defer to @Hilbrand as to which option is best.

Reply:
No, I haven't tried that.
I think rewriting `DLinkHNAPCommunication` could be a good compromise, although the result wouldn't be much different from (5) or (6), imho.

I started implementing it by myself because that's the way I learn. I haven't heard of SOAP or HNAP 10 days ago.
If you don't like my implementation for whatever reason, don't use it. I'm fine with (4). Really. I wrote the code to learn HNAP and SOAP.
It's hard to ask for help when you don't know where to start. A posterior it's easy to point out what could have been asked, but a week ago or even yesterday morning, I hadn't had a clue what to ask precisely.

Reply:
I had a look at the different implementations. The original binding is already some time old and the reviews have become much stricter. So I was going to make a number of review comments only to see some of the code in the pr is copied from the other device implementation.That makes it a bit difficult to give a good suggestion. The code here would improve if some of the communications part is moved out of the handler class in a similar way as the current implementation does (I mean not necessarily similar as in the same implementation, but similar as in what each class does and what interfaces it has). This would make it also easier to see what the overlap is, and how difficult it is to reuse/improve the exisiting implementation, like with `DLinkHNAPCommunication` and if it can be combined. This is basically the first part of option 5. Reviewers can comment, and you can than see what it would take to combine or what would be needed to reuse/adapt existing code to reduce code duplication/complexit or if we leave the combining for a future improvement.